### ISSUE TITLE
Flow 1 alignment disappeared after started training with flow 2

### ISSUE THREAD

**Issue by najmuzsakib11** [NVIDIA/flowtron#91]
I have trained with a private dataset for english language. It is like ljspeech with one speaker. There is no silence at the beginning and end. After training with flow 1 alignment was ok. When I started flow 2 at 300k to 400k attention_weights_0 alignment slowly disappeared but attention_weights_1 alignment was ok.

1. Flow 1 training setup: trained for 2.0M
{
    "train_config": {
        "output_directory": "out_v001",
        "epochs": 10000000,
        "learning_rate": 1e-4,
        "weight_decay": 1e-6,
        "sigma": 1.0,
        "iters_per_checkpoint": 5000,
        "batch_size": 1,
        "seed": 1234,
        "checkpoint_path": "models/flowtron_ljs.pt",
        "ignore_layers": ["speaker_embedding.weight"],
        "include_layers": ["encoder", "embedding"],
        "warmstart_checkpoint_path": "",
        "with_tensorboard": true,
        "fp16_run": false
    },
    "data_config": {
        "training_files": "filelists/ljs_audiopaths_text_sid_train_filelist.txt",
        "validation_files": "filelists/ljs_audiopaths_text_sid_val_filelist.txt",
        "text_cleaners": ["flowtron_cleaners"],
        "p_arpabet": 0.5,
        "cmudict_path": "data/cmudict_dictionary",
        "sampling_rate": 22050,
        "filter_length": 1024,
        "hop_length": 256,
        "win_length": 1024,
        "mel_fmin": 0.0,
        "mel_fmax": 8000.0,
        "max_wav_value": 32768.0
    },
    "dist_config": {
        "dist_backend": "nccl",
        "dist_url": "tcp://localhost:54321"
    },
    "model_config": {
        "n_speakers": 1,
        "n_speaker_dim": 128,
        "n_text": 185,
        "n_text_dim": 512,
        "n_flows": 1,
        "n_mel_channels": 80,
        "n_attn_channels": 640,
        "n_hidden": 1024,
        "n_lstm_layers": 2,
        "mel_encoder_n_hidden": 512,
        "n_components": 0,
        "mean_scale": 0.0,
        "fixed_gaussian": true,
        "dummy_speaker_embedding": false,
        "use_gate_layer": true
    }
}
![Screenshot 2020-12-07 at 2 16 10 PM](https://user-images.githubusercontent.com/47039457/101326074-f44f0000-3896-11eb-99d8-ced201e101b9.png)
![Screenshot 2020-12-07 at 2 15 44 PM](https://user-images.githubusercontent.com/47039457/101326108-ffa22b80-3896-11eb-9926-e3ff99149717.png)

As you can see attention looks ok. But when I started training flow 2 with warmstart and set ignore_layers=None attention disappeared.
2. Flow 2 training setup: trained for 1.2M
{
    "train_config": {
        "output_directory": "out_v002",
        "epochs": 10000000,
        "learning_rate": 1e-4,
        "weight_decay": 1e-6,
        "sigma": 1.0,
        "iters_per_checkpoint": 5000,
        "batch_size": 1,
        "seed": 1234,
        "checkpoint_path": "",
        "ignore_layers": [],
        "include_layers": [],
        "warmstart_checkpoint_path": "out_v001/model_2000000",
        "with_tensorboard": true,
        "fp16_run": false
    },
    "data_config": {
        "training_files": "filelists/ljs_audiopaths_text_sid_train_filelist.txt",
        "validation_files": "filelists/ljs_audiopaths_text_sid_val_filelist.txt",
        "text_cleaners": ["flowtron_cleaners"],
        "p_arpabet": 0.5,
        "cmudict_path": "data/cmudict_dictionary",
        "sampling_rate": 22050,
        "filter_length": 1024,
        "hop_length": 256,
        "win_length": 1024,
        "mel_fmin": 0.0,
        "mel_fmax": 8000.0,
        "max_wav_value": 32768.0
    },
    "dist_config": {
        "dist_backend": "nccl",
        "dist_url": "tcp://localhost:54321"
    },
    "model_config": {
        "n_speakers": 1,
        "n_speaker_dim": 128,
        "n_text": 185,
        "n_text_dim": 512,
        "n_flows": 2,
        "n_mel_channels": 80,
        "n_attn_channels": 640,
        "n_hidden": 1024,
        "n_lstm_layers": 2,
        "mel_encoder_n_hidden": 512,
        "n_components": 0,
        "mean_scale": 0.0,
        "fixed_gaussian": true,
        "dummy_speaker_embedding": false,
        "use_gate_layer": true
    }
}

![Screenshot 2020-12-07 at 1 43 03 PM](https://user-images.githubusercontent.com/47039457/101326280-3bd58c00-3897-11eb-86a8-d3e8343c540a.png)
![Screenshot 2020-12-07 at 1 37 51 PM](https://user-images.githubusercontent.com/47039457/101326302-4728b780-3897-11eb-8701-d280eb1a530e.png)

**Comment by kevjshih** [comment]
This is just a guess but your loss curves look pretty noisy. This might be because you're training at batch size 1. I believe the latest commit adds gradient clipping, which you can try out. Another option is to try lowering the global learning rate, or even just the learning rate of parameters corresponding to those trained before enabling the 2nd step.

**Comment by najmuzsakib11** [comment]
Thank you for the reply. I will try gradient clipping.

**Comment by najmuzsakib11** [comment]
Hi @kevjshih , I have warm started from previous flow=1 checkpoints to train for flow=2. I am using batch_size = 2, Here is my new config:
{
    "train_config": {
        "output_directory": "outdir",
        "epochs": 10000000,
        "optim_algo": "RAdam",
        "learning_rate": 1e-3,
        "weight_decay": 1e-6,
        "grad_clip_val": 1,
        "sigma": 1.0,
        "iters_per_checkpoint": 1000,
        "batch_size": 2,
        "seed": 1234,
        "checkpoint_path": "",
        "ignore_layers": [],
        "finetune_layers": [],
        "include_layers": ["speaker", "encoder", "embedding"],
        "warmstart_checkpoint_path": "",
        "with_tensorboard": true,
        "fp16_run": false
    },
    "data_config": {
        "training_files": "filelists/ljs_audiopaths_text_sid_train_filelist.txt",
        "validation_files": "filelists/ljs_audiopaths_text_sid_val_filelist.txt",
        "text_cleaners": ["flowtron_cleaners"],
        "p_arpabet": 0.5,
        "cmudict_path": "data/cmudict_dictionary",
        "sampling_rate": 22050,
        "filter_length": 1024,
        "hop_length": 256,
        "win_length": 1024,
        "mel_fmin": 0.0,
        "mel_fmax": 8000.0,
        "max_wav_value": 32768.0,
        "use_attn_prior": false,
        "attn_prior_threshold": 1e-4
    },
    "dist_config": {
        "dist_backend": "nccl",
        "dist_url": "tcp://localhost:54321"
    },
    "model_config": {
        "n_speakers": 1,
        "n_speaker_dim": 128,
        "n_text": 185,
        "n_text_dim": 512,
        "n_flows": 2,
        "n_mel_channels": 80,
        "n_attn_channels": 640,
        "n_hidden": 1024,
        "n_lstm_layers": 2,
        "mel_encoder_n_hidden": 512,
        "n_components": 0,
        "mean_scale": 0.0,
        "fixed_gaussian": true,
        "dummy_speaker_embedding": false,
        "use_gate_layer": true,
        "use_cumm_attention": false
    }
}
command: CUDA_VISIBLE_DEVICES=1,0 python -m torch.distributed.launch --use_env --nproc_per_node=2 train.py -c config.json -p train_config.warmstart_checkpoint_path="model_2430000"
I am getting alignment graph bit different then before.
![Screenshot 2020-12-13 at 6 58 58 PM](https://user-images.githubusercontent.com/47039457/102013158-99c31180-3d78-11eb-9dfa-9ecf1e8619b8.png)
Is there any problem? Can you clarify why it is like this?

**Comment by najmuzsakib11** [comment]
I have some farther queries. 
1. Is flow 2 necessary for a single speaker?
2. When to finetune and when to use warmstart? What is the difference?

**Comment by kevjshih** [comment]
The warmstart_checkpoint_path setting in our code mostly refers to checkpoint loading where the architecture doesn't have to match exactly. This is the case for loading a single flow step model for 2 flow steps, where the former is a strict subset of the latter. We also use this when initializing the text encoder weights with those from a pretrained tacotron2 model, where only the text encoder parts should match. I believe the more proper definition of warmstart is to start with a low learning rate and ramp up. We aren't doing anything like that in our code, as far as I recall.

Flow 2 typically adds a significant qualitative improvement, though I wouldn't say it's strictly necessary. 

I noticed your filelists are still pointing to the LJS filelists, but you mentioned you're using a private dataset? I assume you must have overwritten the original contents for this to work.

One more thing you can try is setting use_attn_prior=true in the data_config.

**Comment by najmuzsakib11** [comment]
Hi @kevjshih thank you for the detail answer.  I am getting good results then before with what you suggested. 
Yes i have overwritten the LJS filelists for my private dataset.
ok. I will try use_attn_prior.

**Comment by najmuzsakib11** [comment]
I am trying to train on multi-speaker dataset. I am fine-tuning  Flowtron LibriTTS2K model. I am using different set of speaker_id(not any 123 id of LibriTTS2K model) from my dataset. I am not getting any attention. The attention line looks flat. I am using the same config as before just changed the filelists and n_speaker according to my dataset.
What speaker ids should i use. Do i need to do anything else?
